man/detect-ng.1
.TH DETECT-NG "1" "October 2026" "objtools 1.10" "User Commands"
.SH NAME
detect-ng: display signature information about PE files.

.SH SYNOPSIS
detect-ng [\fB\-h\fR|\fB\-\-help\fR]
          [\fB\-v\fR|\fB\-\-version\fR]
          [\fB\-V\fR|\fB\-\-verbose\fR]
          [\fB\-c\fR|\fB\-\-create\fR]
          [\fB\-e\fR|\fB\-\-ep-only\fR]
          [\fB\-m\fR|\fB\-\-match-once\fR]
          [\fB\-X\fR|\fB\-\-hash\fR]
          [\fB\-E\fR|\fB\-\-entropy\fR]
          pefile...
          sigfile

.SH DESCRIPTION
\fBdetect-ng\fR displays information about one or more PE format object files. The options control what particular information to display.

\fIpefile...\fR are the object files to be examined. 32-bit and 64-bit PE files are supported.

\fIsigfile\fR are the signature file to be applied.

.SH OPTIONS
.IP "\fB\-h\fR" 4
.IX Item "-h"
.PD 0
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
.PD
Print a summary of the options to detect-ng and exit.
.IP "\fB\-v\fR" 4
.IX Item "-v"
.PD 0
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
.PD
Print the version number of detect-ng and exit.
.IP "\fB\-V\fR" 4
.IX Item "-V"
.PD 0
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
.PD
Print additional information.
.IP "\fB\-c\fR" 4
.IX Item "-c"
.PD 0
.IP "\fB\-\-create\fR" 4
.IX Item "--create"
.PD
Create a signature database.
.IP "\fB\-e\fR" 4
.IX Item "-e"
.PD 0
.IP "\fB\-\-ep\-only\fR" 4
.IX Item "--ep-only"
Only consider signatures from entry point.
.IP "\fB\-m\fR" 4
.IX Item "-m"
.PD 0
.IP "\fB\-\-match\-once\fR" 4
.IX Item "--match-once"
.PD
Match only one signature, then stop.
.IP "\fB\-X\fR" 4
.IX Item "-X"
.PD 0
.IP "\fB\-\-hash\fR" 4
.IX Item "--hash"
.PD
Calculates and displays the SHA-256 hash.
.IP "\fB\-E\fR" 4
.IX Item "-E"
.PD 0
.IP "\fB\-\-entropy\fR" 4
.IX Item "--entropy"
.PD
Displays the entropy of the object.

.SH SEE ALSO
detect-ng(1), convert-ng(1), coredump-ng(1), objcopy-ng(1), objdump-ng(1), objdwarf-ng(1), objhash-ng(1), readelf-ng(1), readpe-ng(1)

.SH COPYRIGHT
.IX Header "COPYRIGHT"
MIT License
.PP
Copyright (c) 2025 Kyle Wickens

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.



CHANGELOG
v1.2
-----------------------------------------------------------------------
[+] Added big-endian support.
[+] Added MIPS architecture support.
[+] Improved RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.1
-----------------------------------------------------------------------
[+] Added ARM Thumb architecture support.
[+] Added RISC-V architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################

v1.0
-----------------------------------------------------------------------
[+] Added ARM architecture support.
[+] Improved x86-64 architecture support.
[+] Enhanced readelf-ng and objdump-ng.
[+] Bug fixes.
#######################################################################



src/detect-ng.c
#include "buffer.h"
#include "printf.h"
#include "detect.h"
#include "options.h"
#include "objutils.h"

#include "static/detect-ng.ci"

#define THIS_NAME "detect-ng"

static int usage(poptions_t o, const char* name, const args_t args[], const int ecode) {
  int n = 0;
  n += usage_name(o, name, args, zDESCRIPTIONPE);
  n += usage_synopsis0(o, name, args);
  n += usage_objname(o, name, "pefile");
  n += usage_onename(o, name, "sigfile");
  n += printf_eol();
  n += usage_description(o, name, args, zPEDESCRIPTION);
  n += usage_options0(o, name, args);
  n += usage_seealso(o, name, args);
  n += usage_copyright(o, name, args);

  return ecode;
}

static int get_options_detect(poptions_t o, int argc, char** argv, char* name) {
  if (0 == argc) {
    return usage(o, THIS_NAME, zDETECTARGS, ECODE_ARGUMENTS);
  }

  strname(o->prgname, name);
  printf_errname(o->prgname);

  for (int i = 0; i < argc; ++i) {
    if ('-' == argv[i][0] && '-' == argv[i][1]) {
      imode_t action = get_options2(o, zDETECTARGS, argv[i]);
      if (0 == action) {
        return odeath(o, THIS_NAME, argv[i] + 2);
      }
      o->action |= action;
    } else if ('-' == argv[i][0]) {
      imode_t action = get_options1(o, zDETECTARGS, argv[i]);
      if (0 == action) {
        return odeath(o, THIS_NAME, argv[i] + 1);
      }
      o->action |= action;
    } else if (0 == o->inpname0[0]) {
      xstrncpy(o->inpname0, argv[i], NELEMENTS(o->inpname0));
    } else if (0 == o->inpname1[0]) {
      xstrncpy(o->inpname1, argv[i], NELEMENTS(o->inpname1));
    } else if (0 == o->outname0[0]) {
      xstrncpy(o->outname0, argv[i], NELEMENTS(o->outname0));
    } else {
      return odeath(o, THIS_NAME, argv[i]);
    }
  }

  if (MODE_ISANY(o->action, OPTPROGRAM_VERSION)) {
    return version0(o, THIS_NAME, zDETECTARGS);
  }

  if (MODE_ISANY(o->action, OPTPROGRAM_HELP)) {
    return usage(o, THIS_NAME, zDETECTARGS, ECODE_OK);
  }

  return ECODE_OK;
}

int main(int argc, char* argv[]) {
  int x = ECODE_MALLOC;
  poptions_t o = omalloc();
  if (o) {
    x = get_options_detect(o, argc - 1, argv + 1, argv[0]);
    if (ECODE_ISOK(x) && o->inpname0[0]) {
      pbuffer_t p = bopen(o->inpname0);
      if (p) {
        if (isPE(p)) {
          x = detect(p, o);
        } else if (MODE_ISANY(o->action, OPTDETECT_CREATE)) {
          if (o->outname[0]) {
            x = detect_create(p, o);
          } else {
            printf_e("missing output file.");
          }
        } else {
          printf_e("'%s': invalid file format.", o->inpname0);
        }
      } else {
        printf_e("'%s': no such file.", o->inpname0);
      }

      bfree(p);
    }

    ofree(o);
  }

  return x;
}

